ansible/roles/docker/tasks/main.yml

```diff
@@ -41,26 +42,63 @@
   shell: systemctl status docker-tcp.socket | grep "Active:"
   register: docker_tcp_socket_state
 
+- include: create_docker_device.yml
+  when: docker_device != ""
+
 - name: copy systemd units for docker tcp socket settings
   template: src=docker-tcp.j2 dest=/etc/systemd/system/docker-tcp.socket
   register: docker_tcp_socket
 
   # tcp socket service requires docker service to be started after it
-- name: start docker tcp socket service
-  shell: sudo systemctl stop docker && sudo systemctl start docker-tcp.socket && sudo systemctl start docker
-  when: (docker_tcp_socket | changed) or not docker_tcp_socket_state.stdout | match("Active.* active")
+- name: reload systemd configuration
+  shell: sudo systemctl daemon-reload
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: stop docker
+  service:
+    name: docker
+    state: stopped
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: start docker-tcp service
+  service:
+    name: docker-tcp.socket
+    state: started
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: check docker service state
+  shell: sudo systemctl status docker | grep 'Active.*active' -o
+  ignore_errors: true
+  register: docker_service_state
+  tags:
+    - prebake-for-dev
 
 - name: remove the docker key file, if any. It shall be regenerated by docker on restart
   file: name=/etc/docker/key.json  state=absent
 
 # XXX: service module doesn't do daemon-reload yet, so need to use shell module here
 # https://github.com/ansible/ansible-modules-core/issues/191
-- name: restart docker
+- name: reload docker systemd configuration
   #service: name=docker state=restarted
-  shell: sudo systemctl daemon-reload && sudo systemctl restart docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
   tags:
     - prebake-for-dev
 
+# XXX: this needs to happen twice after setting up the thin provisioner because
+#      of some docker bug I've not investigated.
+- name: restart docker (first time)
+  service:
+    name: docker
+    state: restarted
+  when: thin_provisioned|changed
+  ignore_errors: true
+
+- name: ensure docker is started
+  service:
+    name: docker
+    state: started
```

Comment:
oh. you're right. hm. I will look at this asap.

On 2 Mar 2016, at 13:57, Madhav Puri wrote:

>> tags:
>>   - prebake-for-dev
>>
>> +# XXX: this needs to happen twice after setting up the thin 
>> provisioner because
>> +#      of some docker bug I've not investigated.
>> +- name: restart docker (first time)
>> +  service:
>> +    name: docker
>> +    state: restarted
>> +  when: thin_provisioned|changed
>> +  ignore_errors: true
>> +
>> +- name: ensure docker is started
>> +  service:
>> +    name: docker
>> +    state: started
>
> @erikh
>
>> started also means to enable it on boot
>
> for service to be enabled we need to set the `enabled` attribute in 
> `service` module. But I am ok if you want to leave this task as is.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/contiv/volplugin/pull/152/files#r54798917